references/segmentation/presets.py

```diff
@@ -1,39 +1,42 @@
 import torch
-import transforms as T
+from torchvision.prototype import features, transforms as T
+from torchvision.prototype.transforms import functional as F
+from transforms import RandomCrop
 
 
-class SegmentationPresetTrain:
-    def __init__(self, *, base_size, crop_size, hflip_prob=0.5, mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225)):
-        min_size = int(0.5 * base_size)
-        max_size = int(2.0 * base_size)
+class WrapIntoFeatures(T.Transform):
+    def forward(self, sample):
+        image, mask = sample
+        return image, features.Mask(F.to_image_tensor(mask).squeeze(0), dtype=torch.int64)
+
 
-        trans = [T.RandomResize(min_size, max_size)]
+class SegmentationPresetTrain(T.Compose):
+    def __init__(self, *, base_size, crop_size, hflip_prob=0.5, mean=(0.485, 0.456, 0.406), std=(0.229, 0.224, 0.225)):
+        transforms = [
+            WrapIntoFeatures(),
+            T.RandomResize(min_size=int(0.5 * base_size), max_size=int(2.0 * base_size)),
+        ]
         if hflip_prob > 0:
-            trans.append(T.RandomHorizontalFlip(hflip_prob))
-        trans.extend(
+            transforms.append(T.RandomHorizontalFlip(hflip_prob))
+        transforms.extend(
             [
-                T.RandomCrop(crop_size),
-                T.PILToTensor(),
+                RandomCrop(crop_size, pad_if_needed=True),
```

Comment:
@pmeier I think that's not equivalent with before.

Previously we padded as follows:
https://github.com/pytorch/vision/blob/d367a01a18a3ae6bee13d8be3b63fd6a581ea46f/references/segmentation/transforms.py#L9-L16

Now we pad differently:
https://github.com/pytorch/vision/blob/08743385d93a0aa4145da6f6db49bfa00df148a3/torchvision/prototype/transforms/_geometry.py#L438-L444

This inconsistency could be the reason we get accuracy issues.

Reply:
Here is how the padding is done:

1. references: perform a minimum bottom-right pad. Meaning, only pad the minimum amount so that `crop` doesn't complain. If both dimensions need to be padded, the `RandomCrop` transform is a no-op.

2. transforms (stable or prototype): perform a pad on all directions with twice the minimum amount. For example, an image of size `(16, 16)` with a crop size of `(18, 18)` get padded with `2 * (18 - 16) == 4` on all sides to `(24, 24)` and only then the crop will be performed. This behavior is intentional, see #564, to always enable random cropping even if there is padding.

I don't think we can unify both transforms in a way that supports both behaviors. My suggestion is to add `_compute_padding(...)` method on `transforms.RandomCrop` that we can overwrite in the references.

Thoughts? @datumbox @vfdev-5

Reply:
@pmeier Thanks for your debugging, I agree with your findings. This change of behaviour is likely to be causing the accuracy regression reported above.

In an offline chat, @vfdev-5 suggested decomposing Pad and Crop to achieve a similar effect, which is a great idea. Your proposal to overwrite the `RandomCrop` in references is a viable approach as well. As you both seem to suggest, we should avoid making the `transforms.RandomCrop` more complicated than it is.

Reply:
I've introduced a `PadIfSmaller` transform in e912976a5b30018fda5cc50d9169fc1b697ac84b as discussed in #6591.